### DISCUSSION TITLE
How to make json_encoders work for nested custom models ?

### DISCUSSION THREAD

**Discussion by Shraneid** [pydantic/pydantic#5041] [Question] [answered]
Hello,
I have a BaseModel object nested into another but it seems like my json_encoder is not working properly, I'm not sure if I'm misunderstanding something or if it's a bug

here's a simple example

```
class MyClass(BaseModel):
    text: str

class WithCustomEncoders(BaseModel):
    myC: MyClass

    class Config:
        json_encoders = {
            MyClass: lambda v: "blablablablablabalba",
        }

m = WithCustomEncoders(myC={"text": "test"})
```

is this case, `m` is getting parsed as expected to `m --> {myC=MyClass(text='test')}`

but `m.json()` seems to only return  the dict result and overlook the encoder
`m.json() --> '{"myC": {"text": "test"}}'`

I would expect `m.json() --> '{"myC": "blablablablablabalba"}'`

it's not the same as https://github.com/pydantic/pydantic/issues/1671 since my class is getting parsed correctly
it's not like https://stackoverflow.com/questions/71451320/pydantic-json-encoders-for-builtin-types-float-int-etc since I don't use built-in types

I'm a bit lost, any help appreciated :)

**Comment by dmontagu** [comment]
This _looks_ like a bug to me. In this section, there is a comment explicitly calling out why it is doing what it is doing during the process of json encoding, but if I look at the value of `data` on the final line above in a debugger during the `.json()` call, I can see it has already been converted to dicts recursively:

```
data = {'myC': {'text': 'test'}}
```

This is why it is not getting encoded properly -- it has already been converted to dicts.

https://github.com/pydantic/pydantic/blob/ce45f6ebcb60abf54646376ec8bad0dc17df43bc/pydantic/main.py#L489-L505

@samuelcolvin I am surprised there isn't a test for this, so I'm wondering if maybe there is some trick here. Or maybe it's just a bug?

**Comment by dmontagu** [comment, marked as answer]
Oh, I looked in the tests and figured it out, I think.

@Shraneid try using `.json(models_as_dict=False)` instead of just `.json()`. It worked for me with your code. 👍

**Reply by Shraneid** [reply]
Oh thanks a lot !
Can't look at it now but will def try it tomorrow
Weird that it isn't mentioned on the wiki in the json_encoders part, would've saved me quite some time 😆

**Reply by Shraneid** [reply]
Coming back, I've run a few tests and everything seems to be working perfectly ! Thanks a lot for the help !

Do you know if I can PR the wiki somehow to add a line about this feature ?
(btw, had to battle a bit since I was on 1.8.x and it released in 1.9.0)
